src/core/cluster_analysis/Cluster.hpp

```diff
@@ -32,6 +35,8 @@ namespace ClusterAnalysis {
 /** @brief Represents a single cluster of particles */
 class Cluster {
 public:
+  explicit Cluster(std::weak_ptr<BoxGeometry const> const &box_geo)
```

Comment:
The `BoxGeometry` should be owned by the `System` class, like most other pointer members of `System`. Objects of type `Cluster` may outlive the `System` class in Python. I don't think I left any `std::shared_ptr<BoxGeometry>` anywhere in the code, except for the `System` class. While it probably wouldn't hurt to share ownership of `BoxGeometry` (it's just a collection of arrays), for other classes it's not safe for them to outlive the `System` class (we have an exception mechanism to generate a meaningful error message in that case). I gave all system-owned pointers the same treatment with regards to lifetime: it can only be extended by client code like `Cluster` for the short duration of an operation.

Sometimes I have to store a `std::shared_ptr<X>` in a class. This is for example the case of Ekin and P3M, where an owning pointer to `boost::mpi::environment` is stored to guarantee we can deallocate the FFT plans even if the `System` class has expired. FFT plans deallocation requires the boost MPI communicator, whose lifetime is tied to the MPI environment. Similar argument for the CUDA handle, needed to `deallocate` thrust vectors stored in device memory.